src/internal/permissions.ts

```diff
@@ -0,0 +1,429 @@
+  assetAddress: string;
+  capabilities: Record<NFTCapability, boolean>;
+}
+
+export interface NFTCollectionPermission {
+  type: PermissionType.NFTCollection;
+  collectionAddress: string;
+  capabilities: Record<NFTCollectionCapability, boolean>;
+}
+
+export enum NFTCollectionCapability {
+  transfer = "transfer",
+  mutate = "mutate",
+}
+
+export interface GenericPermission {
+  type: Exclude<string, PermissionType | `${PermissionType}`>;
+  [key: string]: string | Exclude<string, PermissionType | `${PermissionType}`>;
+}
+
+// revoke permission types
+export type RevokeFungibleAssetPermission = Pick<FungibleAssetPermission, "type" | "asset">;
+export type RevokeNFTPermission = Pick<NFTPermission, "type" | "assetAddress">;
+
+// factories -- on second thought, i think these are dumb
+export function FungibleAssetPermission(args: Omit<FungibleAssetPermission, "type">): FungibleAssetPermission {
+  return { type: PermissionType.FungibleAsset, ...args };
+}
+export function GasPermission(args: Omit<GasPermission, "type">): GasPermission {
+  return { type: PermissionType.Gas, ...args };
+}
+export function NFTPermission(args: Omit<NFTPermission, "type">): NFTPermission {
+  return { type: PermissionType.NFT, ...args };
+}
+export function NFTCollectionPermission(args: Omit<NFTCollectionPermission, "type">): NFTCollectionPermission {
+  return { type: PermissionType.NFTCollection, ...args };
+}
+
+// revoke factories
+export function RevokeFungibleAssetPermission(args: Omit<RevokeFungibleAssetPermission, "type">): RevokeFungibleAssetPermission {
+  return { type: PermissionType.FungibleAsset, ...args };
+}
+export function RevokeNFTPermission(args: Omit<RevokeNFTPermission, "type">): RevokeNFTPermission {
+  return { type: PermissionType.NFT, ...args };
+}
+
+// mostly for testing
+export function GenericPermission(args: GenericPermission): GenericPermission {
+  return {...args}
+}
+
+// functions
+
+export interface PermissionTemp{
+  asset: string;
+  type: string;
+  remaining: string;
+}
+export async function getPermissions<T extends PermissionType>({
+  aptosConfig,
+  primaryAccount,
+  subAccount,
+  filter
+}: {
+  aptosConfig: AptosConfig;
+  primaryAccount: SingleKeyAccount;
+  subAccount: AbstractedEd25519Account;
+  filter?: T;
+}): Promise<FilteredPermissions<T>> {
+  const handle = await getHandleAddress({ aptosConfig, primaryAccount, subAccount });
+
+  const res = await fetch(`http://127.0.0.1:8080/v1/accounts/${handle}/resources`);
+
+  type NodeDataResponse = Array<{
+    type: string;
+    data: {
+      perms: {
+        data: Array<{
+          key: { data: string; type_name: string };
+          value: string;
+        }>;
+      };
+    };
+  }>;
+
+  const data = (await res.json()) as NodeDataResponse;
+
+  const permissions = data[0].data.perms.data.map((d) => {
+    switch (d.key.type_name) {
+      case "0x1::fungible_asset::WithdrawPermission": // can this be better? i dont rly like this
```

Comment:
Replace with the capability enum. What don't you like about it? The hardcoded string? The potential for a large `switch`?

Reply:
yea feels hardcoded but NBD if no way around it